baremetal/machine_manager_test.go

```diff
@@ -0,0 +1,1697 @@
+
+	if err != nil {
+		log.Printf("AddToScheme failed: %v", err)
+	}
+
+	host1 := bmh.BareMetalHost{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "host1",
+			Namespace: "myns",
+		},
+		Spec: bmh.BareMetalHostSpec{
+			ConsumerRef: &corev1.ObjectReference{
+				Name:       "someothermachine",
+				Namespace:  "myns",
+				Kind:       "Machine",
+				APIVersion: capi.GroupVersion.String(),
+			},
+		},
+	}
+	host2 := bmh.BareMetalHost{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "host2",
+			Namespace: "myns",
+		},
+	}
+	host3 := bmh.BareMetalHost{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "host3",
+			Namespace: "myns",
+		},
+		Spec: bmh.BareMetalHostSpec{
+			ConsumerRef: &corev1.ObjectReference{
+				Name:       "machine1",
+				Namespace:  "myns",
+				Kind:       "Machine",
+				APIVersion: capi.GroupVersion.String(),
+			},
+		},
+	}
+	host4 := bmh.BareMetalHost{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "host4",
+			Namespace: "someotherns",
+		},
+	}
+	discoveredHost := bmh.BareMetalHost{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "discoveredHost",
+			Namespace: "myns",
+		},
+		Status: bmh.BareMetalHostStatus{
+			ErrorMessage: "this host is discovered and not usable",
+		},
+	}
+	hostWithLabel := bmh.BareMetalHost{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "hostWithLabel",
+			Namespace: "myns",
+			Labels:    map[string]string{"key1": "value1"},
+		},
+	}
+
+	bmmconfig, infrastructureRef := newConfig(t, "", map[string]string{},
+		[]capbm.HostSelectorRequirement{})
+	bmmconfig2, infrastructureRef2 := newConfig(t, "", map[string]string{"key1": "value1"},
+		[]capbm.HostSelectorRequirement{})
+	bmmconfig3, infrastructureRef3 := newConfig(t, "", map[string]string{"boguskey": "value"},
+		[]capbm.HostSelectorRequirement{})
+	bmmconfig4, infrastructureRef4 := newConfig(t, "", map[string]string{},
+		[]capbm.HostSelectorRequirement{
+			capbm.HostSelectorRequirement{
+				Key:      "key1",
+				Operator: "in",
+				Values:   []string{"abc", "value1", "123"},
+			},
+		})
+	bmmconfig5, infrastructureRef5 := newConfig(t, "", map[string]string{},
+		[]capbm.HostSelectorRequirement{
+			capbm.HostSelectorRequirement{
+				Key:      "key1",
+				Operator: "pancakes",
+				Values:   []string{"abc", "value1", "123"},
+			},
+		})
+
+	testCases := map[string]struct {
+		Machine          capi.Machine
+		Hosts            []runtime.Object
+		BMMachine        *capbm.BareMetalMachine
+		ExpectedHostName string
+	}{
+		"Pick host2 which lacks ConsumerRef": {
+			Machine: capi.Machine{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "machine1",
+					Namespace: "myns",
+				},
+				TypeMeta: metav1.TypeMeta{
+					Kind:       "Machine",
+					APIVersion: capi.GroupVersion.String(),
+				},
+				Spec: capi.MachineSpec{
+					InfrastructureRef: *infrastructureRef2,
+				},
+			},
+			Hosts:            []runtime.Object{&host2, &host1},
+			BMMachine:        bmmconfig2,
+			ExpectedHostName: host2.Name,
```

Comment:
Could you create a function per type of object that returns the object you need (parameterized to offer flexibility). That would remove a big amount of duplicated code in this whole file. You can check the controllers tests suite that has newMachine and newBaremetalMachine and newCluster and newBaremetalCluster that might even be directly reusable.

Reply:
major part done.
some structures (less generic use) still not populated by functions, this is a conscious decision.